services/packages/arch/service.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package arch
+
+import (
+	"bytes"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	"code.gitea.io/gitea/models/db"
+	pkg_model "code.gitea.io/gitea/models/packages"
+	repository "code.gitea.io/gitea/models/repo"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/packages"
+	"code.gitea.io/gitea/modules/packages/arch"
+	"code.gitea.io/gitea/modules/storage"
+)
+
+// Get data related to provided filename and distribution, for package files
+// update download counter.
+func GetPackageFile(ctx *context.Context, distro, file string) (storage.Object, error) {
+	pkgfile := &pkg_model.PackageFile{CompositeKey: distro + "-" + file}
+
+	ok, err := db.GetEngine(ctx).Get(pkgfile)
+	if err != nil || !ok {
+		return nil, fmt.Errorf("%+v %t", err, ok)
+	}
+
+	b, err := pkg_model.GetBlobByID(ctx, pkgfile.BlobID)
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.HasSuffix(file, ".pkg.tar.zst") {
+		err = pkg_model.IncrementDownloadCounter(ctx, pkgfile.VersionID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return packages.NewContentStore().Get(packages.BlobHash256Key(b.HashSHA256))
+}
```

Comment:
Why is there no version parameter? This method should use `package_service.GetPackageFileStream` or be replaced with one of the other helper methods.

Reply:
@KN4CK3R 

Arch package files contain version, release number and architecture in file name. For example:

```
zsh-5.9-4-x86_64.pkg.tar.zst
zsh-autosuggestions-0.7.0-2-any.pkg.tar.zst
zsh-syntax-highlighting-0.7.1-1-any.pkg.tar.zst
```

I have 2 options to get package file property from database:

- Using existing helper methods:
```go
func GetPackageFile(ctx *context.Context, distro, file string) (io.ReadSeekCloser, error) {
	var (
		splt     = strings.Split(file, "-")
		pkgname  = strings.Join(splt[0:len(splt)-3], "-")
		vername  = splt[len(splt)-3] + "-" + splt[len(splt)-2]
		filename = strings.TrimSuffix(file, ".sig")
		filekey  = distro + "-" + filename
	)

	version, err := pkg_model.GetVersionByNameAndVersion(
		ctx, ctx.Package.Owner.ID, pkg_model.TypeArch, pkgname, vername,
	)
	if err != nil {
		return nil, err
	}

	pkgfile, err := pkg_model.GetFileForVersionByName(
		ctx, version.ID, filename, filekey,
	)
	if err != nil {
		return nil, err
	}

	filestream, _, _, err := pkg_service.GetPackageFileStream(ctx, pkgfile)
	return filestream, err
}
```

- Using direct call to database with package composite key:
```go
func GetPackageFile(ctx *context.Context, distro, file string) (io.ReadSeekCloser, error) {
	pkgfile := &pkg_model.PackageFile{CompositeKey: distro + "-" + file}

	ok, err := db.GetEngine(ctx).Get(pkgfile)
	if err != nil || !ok {
		return nil, fmt.Errorf("%+v %t", err, ok)
	}

	filestream, _, _, err := pkg_service.GetPackageFileStream(ctx, pkgfile)
	return filestream, err
}
```

Which one would be better for current scenario?

Reply:
Added function that gets package file property from file name. It is possible, since file name contains all required information.